src/Datadog.Trace/Sampling/CustomSamplingRule.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Datadog.Trace.Logging;
+using Newtonsoft.Json;
+
+namespace Datadog.Trace.Sampling
+{
+    internal class CustomSamplingRule : ISamplingRule
+    {
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<CustomSamplingRule>();
+        private static readonly TimeSpan RegexTimeout = TimeSpan.FromSeconds(1);
+
+        private readonly float _samplingRate;
+        private readonly string _serviceNameRegex;
+        private readonly string _operationNameRegex;
+
+        private bool _hasPoisonedRegex = false;
+
+        public CustomSamplingRule(
+            float rate,
+            string ruleName,
```

Comment:
what is `ruleName` used for?

Reply:
feels unnecessary?

Reply:
Was for testing and more easily identifying rules.
When logging, the name is logged with the rule. 
If it is not specified, it is given a rule based on it's index in configuration.
`return new CustomSamplingRule(r.SampleRate, r.RuleName ?? $"config-rule-{index}", r.Service, r.OperationName)`

First rule is "config-rule-1", second rule is "config-rule-2", etc.

Reply:
Then it makes it easier to look at config and know which rule a trace is matched on.